Enabling candidate generation with both `non_linear_constraints` and `fixed_features`

jduerholt opened the pull request:
<!--
Thank you for sending the PR! We appreciate you spending the time to make BoTorch better.

Help us understand your motivation by explaining why you decided to make this change.

You can learn more about contributing to BoTorch here: https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md
-->

## Motivation

This PR adds the possibility to use `fixed_features` in combination with `nonlinear_inequality_constraints` as discussed in issue https://github.com/pytorch/botorch/issues/1904.

@Balandat: concerning the solution that you suggested: I think it is much easier as the `f_np_wrapper` is also used for the nonlinear constraints: https://github.com/pytorch/botorch/blob/38e20275188f4cb2b99c8631bb91450b59140308/botorch/generation/gen.py#L224 and `fix_features` is already applied there: https://github.com/pytorch/botorch/blob/38e20275188f4cb2b99c8631bb91450b59140308/botorch/generation/gen.py#L189

So no need for another wrapper, or?

A wrapper function would be only needed if one wants to have the possibility to use the nonlinear constraints also with the reduced domain, then one would need a method, which takes the reduced X as input and augments it with the fixed features. 

Here is some example code, that shows that it works:

``` python
import torch
from botorch.models import SingleTaskGP
from botorch.fit import fit_gpytorch_mll
from botorch.utils import standardize
from gpytorch.mlls import ExactMarginalLogLikelihood
from botorch.utils.sampling import get_polytope_samples

def constraint(X):
    return torch.sum(X[...,:3], dim=-1) -1.80

train_X = torch.rand(10, 3)
Y = 1 - torch.norm(train_X[:,:2] - 0.5, dim=-1, keepdim=True)
Y = Y + 0.1 * torch.randn_like(Y)  # add some noise
train_Y = standardize(Y)

gp = SingleTaskGP(train_X, train_Y)
mll = ExactMarginalLogLikelihood(gp.likelihood, gp)
fit_gpytorch_mll(mll)

from botorch.acquisition import UpperConfidenceBound

UCB = UpperConfidenceBound(gp, beta=0.1)

from botorch.optim import optimize_acqf


bounds = torch.stack([torch.zeros(3), torch.ones(3)])

batch_initial_conditions = get_polytope_samples(
    n=5,
    bounds=bounds,
    inequality_constraints=[(torch.tensor([0,1,2]),torch.tensor([1.,1.,1.]),1.8)],
    n_burnin=1000
).unsqueeze(-2)

candidate, acq_value = optimize_acqf(
    UCB, 
    bounds=bounds, 
    q=1, 
    num_restarts=5, 
    raw_samples=20,
    fixed_features={2:0.5}, 
    nonlinear_inequality_constraints = [constraint],
    batch_initial_conditions = batch_initial_conditions
)
print(candidate)  # tensor([0.4887, 0.5063])
print(candidate.sum())
```

### Have you read the [Contributing Guidelines on pull requests](https://github.com/pytorch/botorch/blob/main/CONTRIBUTING.md#pull-requests)?

Yes.

## Test Plan

Unit tests.

Balandat commented on test/optim/test_optimize.py:
```diff
@@ -862,6 +862,26 @@ def nlc4(x):
+            self.assertTrue(
+                torch.allclose(
+                    candidates,
+                    torch.tensor([[2, 1, 1]], **tkwargs),
```
meganit - let's indicate directly that these are expected to be float (even if thats achieved via the kwargs)
```suggestion
                    torch.tensor([[2.0, 1.0, 1.0]], **tkwargs),
```

Balandat commented on test/optim/test_optimize.py:
```diff
@@ -862,6 +862,26 @@ def nlc4(x):
+                nonlinear_inequality_constraints=[nlc1, nlc2, nlc3, nlc4],
+                batch_initial_conditions=batch_initial_conditions,
+                num_restarts=num_restarts,
+                fixed_features={0: 2},
```
Hmm so with this setup this test may pass even if fixing the features doesn't work (given that it only narrows this to a specific solution of the previous solution set). Can you use a different value here? I might also make sense to use a different mock acquisition function so that the optimizer is unique...

jduerholt replied:
I will have a deeper look to the test. All this mocking is sometimes a bit hard for me to understand ;)

Balandat replied:
I don't think there is actually any mocking going on here - it's just that `mock_acq_function` is a `SquaredAcquisitionFunction` rather than a true acquisition function.

If you don't want to change that, one thing you could do would be to just replace one of the constraints to result in a different optimizer.

Balandat reviewed (commented):
Yeah I think this makes sense. If there are multiple fixed features it's going to be slower than optimizing over the reduced domain since the optimizer needs to (i) identify the fixed features and (ii) SLSQP scales pretty poorly with the problem dimension anyway.

But that's probably fine for now (maybe add this comment in code for future reference?

> tensor([0.4887, 0.5063])
 
This doesn't seem to be the right output? Should be 3dim with the third feature fixed to 0.5?

Balandat commented:
Test failure seems unrelated I'll take a look at that

jduerholt commented:
> Yeah I think this makes sense. If there are multiple fixed features it's going to be slower than optimizing over the reduced domain since the optimizer needs to (i) identify the fixed features and (ii) SLSQP scales pretty poorly with the problem dimension anyway.

I am already working on the proper solution with the reduced domain. I will give it a few more minutes and if I think it is feasible without taking too much time, I will add it to the PR.


> This doesn't seem to be the right output? Should be 3dim with the third feature fixed to 0.5?

You care correct, this is copied from the botorch website minimal example 
![image](https://github.com/pytorch/botorch/assets/28678090/35092593-314c-4ee3-91b5-b27bfa96f5c5)
and I forgot to remove it :D

jduerholt commented:
Hi @Balandat,

I implemented now also the solution using the reduced domain. Which one do you prefer?

Best,

Johannes

Balandat commented on botorch/optim/parameter_constraints.py:
```diff
@@ -312,13 +312,48 @@ def _make_linear_constraints(
+            selector.append(idx_X)
+            idx_X += 1
+
+    values = torch.tensor([v for v in fixed_features.values()], dtype=torch.double)
```
```suggestion
    values = torch.tensor(list(fixed_features.values()), dtype=torch.double)
```

Balandat commented on botorch/optim/parameter_constraints.py:
```diff
@@ -312,13 +312,48 @@ def _make_linear_constraints(
+
+        return new_nlc
+
+    for nlc in constraints:
+        new_constraints.append(_wrap_nlc(nlc=nlc))
+    return new_constraints
```
Could just do this
```suggestion
    return [_wrap_nlc(nlc=nlc) for nlc in constraints]
```

Balandat commented on test/optim/test_parameter_constraints.py:
```diff
@@ -215,6 +216,42 @@ def test_make_scipy_linear_constraints_unsupported(self):
+            return x[..., 0] - 1
+
+        # first test with one constraint
+        new_constraints = _generate_unfixed_nonlin_constraints(
+            constraints=[nlc1], fixed_features={1: 2.0}, dimension=3
+        )
+        self.assertAllClose(
+            nlc1(torch.tensor([[4.0, 2.0, 2.0]], device=self.device)),
+            new_constraints[0](torch.tensor([[4.0, 2.0]], device=self.device)),
+        )
```
```suggestion
        new_nlc1,  = _generate_unfixed_nonlin_constraints(
            constraints=[nlc1], fixed_features={1: 2.0}, dimension=3
        )
        self.assertAllClose(
            nlc1(torch.tensor([[4.0, 2.0, 2.0]], device=self.device)),
            new_nlc1(torch.tensor([[4.0, 2.0]], device=self.device)),
        )
```

Balandat commented on test/optim/test_parameter_constraints.py:
```diff
@@ -215,6 +216,42 @@ def test_make_scipy_linear_constraints_unsupported(self):
+        new_constraints = _generate_unfixed_nonlin_constraints(
+            constraints=constraints, fixed_features={1: 2.0}, dimension=3
+        )
+        for i in range(2):
+            self.assertAllClose(
+                constraints[i](torch.tensor([[4.0, 2.0, 2.0]], device=self.device)),
+                new_constraints[i](torch.tensor([[4.0, 2.0]], device=self.device)),
+            )
```
```suggestion
        for nlc, new_nlc in zip(constraints, new_constraints):
            self.assertAllClose(
               nlc(torch.tensor([[4.0, 2.0, 2.0]], device=self.device)),
                new_nlc(torch.tensor([[4.0, 2.0]], device=self.device)),
            )
```

Balandat commented on test/optim/test_parameter_constraints.py:
```diff
@@ -215,6 +216,42 @@ def test_make_scipy_linear_constraints_unsupported(self):
+        new_constraints = _generate_unfixed_nonlin_constraints(
+            constraints=constraints, fixed_features={1: 2.0, 2: 1.0}, dimension=3
+        )
+        for i in range(2):
+            self.assertAllClose(
+                constraints[i](torch.tensor([[4.0, 2.0, 1.0]], device=self.device)),
+                new_constraints[i](torch.tensor([[4.0]], device=self.device)),
+            )
+
```
```suggestion
        for nlc, new_nlc in zip(constraints, new_constraints):
            self.assertAllClose(
                nlc(torch.tensor([[4.0, 2.0, 1.0]], device=self.device)),
                new_nlc(torch.tensor([[4.0]], device=self.device)),
            )

```

Balandat reviewed (approved):
Looks great, just a couple of minor style nits. Thanks for putting this up!

jduerholt commented:
@Balandat, I incorporated your changes and added another test which was still missing (from my perspective). From my side, this is now ready for merge.

Balandat commented:
Great, thanks you. I'll pull it in and merge it.

SebastianAment commented:
> Test failure seems unrelated I'll take a look at that

This was solved by #1919, should pass upon re-running the tests.

SebastianAment commented on botorch/optim/parameter_constraints.py:
```diff
@@ -312,13 +312,44 @@ def _make_linear_constraints(
+    constraints: Optional[List[Callable]],
+    fixed_features: Dict[int, float],
+    dimension: int,
+) -> Optional[List[Callable]]:
```
Two subjective suggestions to make this quicker to parse:
1) How about `_gen_nonlin_constraints_of_variable_features`? Could apply same convention to the linear case below. (cc @Balandat)
2) Could you add a doc-string here, something like "Given a dictionary of `fixed_features`, returns a list of non-linear constraints on the variable (non-fixed) features."

SebastianAment replied:
Thank you! If you're up for it, you can do both in this commit. The only mentions of `_generate_unfixed_lin_constraints` are in:
- `generation/utils.py`,
- `optim/parameter_constraints.py`, and
- `test/optim/parameter_constraints.py`.
A quick search and replace in these files should do it.

jduerholt replied:
Hmm, but do you really prefer the new name? What do you do not like about `_generate_unfixed_lin_constraints`?

jduerholt replied:
But ofc, I can do it ;)

SebastianAment replied:
`_generate_unfixed_(non)lin_constraints` makes it sound like the constraints are "unfixed", whereas it is the features that are. But let's keep it then, especially since the docstring makes it clear now. Thanks for adding it!

SebastianAment commented on botorch/optim/parameter_constraints.py:
```diff
@@ -312,13 +312,44 @@ def _make_linear_constraints(
+
+    values = torch.tensor(list(fixed_features.values()), dtype=torch.double)
+
+    def _wrap_nlc(nlc: Callable) -> Callable:
```
Complete type signature and more explicit naming.
```suggestion
    def _wrap_nonlin_constraint(constraint: Callable[[Tensor], Tensor]) -> Callable[[Tensor], Tensor]:
```

SebastianAment commented on botorch/optim/parameter_constraints.py:
```diff
@@ -312,13 +312,44 @@ def _make_linear_constraints(
+    values = torch.tensor(list(fixed_features.values()), dtype=torch.double)
+
+    def _wrap_nlc(nlc: Callable) -> Callable:
+        def new_nlc(X: Tensor) -> Tensor:
```
```suggestion
        def new_nonlin_constraint(X: Tensor) -> Tensor:
```

SebastianAment commented on botorch/optim/parameter_constraints.py:
```diff
@@ -312,13 +312,44 @@ def _make_linear_constraints(
+        def new_nlc(X: Tensor) -> Tensor:
+            ivalues = values.to(X).expand(*X.shape[:-1], len(fixed_features))
+            X_perm = torch.cat([X, ivalues], dim=-1)
+            return nlc(X_perm[..., selector])
```
```suggestion
            return constraint(X_perm[..., selector])
```

SebastianAment commented on botorch/optim/parameter_constraints.py:
```diff
@@ -312,13 +312,44 @@ def _make_linear_constraints(
+            X_perm = torch.cat([X, ivalues], dim=-1)
+            return nlc(X_perm[..., selector])
+
+        return new_nlc
```
```suggestion
        return new_nonlin_constraint
```

SebastianAment commented on botorch/optim/parameter_constraints.py:
```diff
@@ -312,13 +312,44 @@ def _make_linear_constraints(
+
+        return new_nlc
+
+    return [_wrap_nlc(nlc=nlc) for nlc in constraints]
```
```suggestion
    return [_wrap_nonlin_constraint(constraint=constraint) for constraint in constraints]
```

SebastianAment commented on botorch/optim/parameter_constraints.py:
```diff
@@ -312,13 +312,44 @@ def _make_linear_constraints(
+    constraints: Optional[List[Callable]],
+    fixed_features: Dict[int, float],
+    dimension: int,
+) -> Optional[List[Callable]]:
```
```suggestion
) -> Optional[List[Callable[[Tensor], Tensor]]:
```

SebastianAment commented on botorch/optim/parameter_constraints.py:
```diff
@@ -312,13 +312,44 @@ def _make_linear_constraints(
 
 
+def _generate_unfixed_nonlin_constraints(
+    constraints: Optional[List[Callable]],
```
```suggestion
    constraints: Optional[List[Callable[[Tensor], Tensor]],
```

SebastianAment commented on botorch/generation/utils.py:
```diff
@@ -73,6 +77,7 @@ def _remove_fixed_features_from_optimization(
     upper_bounds: Optional[Union[float, Tensor]],
     inequality_constraints: Optional[List[Tuple[Tensor, Tensor, float]]],
     equality_constraints: Optional[List[Tuple[Tensor, Tensor, float]]],
+    nonlinear_inequality_constraints: Optional[List[Callable]],
```
```suggestion
    nonlinear_inequality_constraints: Optional[List[Callable[[Tensor], Tensor],
```

SebastianAment commented on botorch/generation/utils.py:
```diff
@@ -63,6 +66,7 @@ class _NoFixedFeatures:
     upper_bounds: Optional[Union[float, Tensor]]
     inequality_constraints: Optional[List[Tuple[Tensor, Tensor, float]]]
     equality_constraints: Optional[List[Tuple[Tensor, Tensor, float]]]
+    nonlinear_inequality_constraints: Optional[List[Callable]]
```
```suggestion
    nonlinear_inequality_constraints: Optional[List[Callable[[Tensor], Tensor]
```

SebastianAment reviewed (commented):
Thanks for adding this @jduerholt! Mainly left cosmetic suggestions that will aid readability, as well as completed type signatures. 

Last, could you rename the PR something a little more descriptive like "Enabling candidate generation with both `non_linear_constraints` and `fixed_features`"? This will make it easier to understand what's going on when browsing the commit history on a high level.
